# Exon Alignments

This is a very simplified implementation of the algorithm I proposed.

# Instructions

Currently I'm just comparing a human transcript with the ortholog mouse transcripts.

Currently it takes fasta files as input, which will soon be migrated to a database.

Please set the value of the variable human_transcript_path to the path to a human transcript fasta file

Sorry for the inconvenience, if this algorithm turns out to be working good, we'll wrap it up in a convenient manner.

Run the script as:

```
python3 exon_alignment.py
```

# Requirements

I've written this code in Python3. I've used the following modules:

```
BioPython
```


